vm/src/obj/objset.rs

```diff
@@ -56,536 +55,688 @@ impl PyValue for PyFrozenSet {
+    fn iter(&self, vm: &VirtualMachine) -> PyListIterator {
+        let items = self.elements.values().cloned().collect();
+        let set_list = vm.ctx.new_list(items);
+        PyListIterator {
+            position: Cell::new(0),
+            list: set_list.downcast().unwrap(),
+        }
+    }
 
-    let elements = get_elements(o);
-    let s = if elements.is_empty() {
-        "set()".to_string()
-    } else if let Some(_guard) = ReprGuard::enter(o) {
+    fn repr(&self, vm: &VirtualMachine) -> PyResult<String> {
         let mut str_parts = vec![];
-        for elem in elements.values() {
+        for elem in self.elements.values() {
             let part = vm.to_repr(elem)?;
             str_parts.push(part.value.clone());
         }
 
-        format!("{{{}}}", str_parts.join(", "))
-    } else {
-        "set(...)".to_string()
-    };
-    Ok(vm.new_str(s))
-}
+        Ok(format!("{{{}}}", str_parts.join(", ")))
+    }
 
-pub fn set_contains(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
-    arg_check!(vm, args, required = [(set, None), (needle, None)]);
-    validate_set_or_frozenset(vm, set.class())?;
-    for element in get_elements(set).iter() {
-        match vm._eq(needle.clone(), element.1.clone()) {
-            Ok(value) => {
-                if objbool::get_value(&value) {
-                    return Ok(vm.new_bool(true));
+    fn add(&mut self, item: &PyObjectRef, vm: &VirtualMachine) -> PyResult {
+        insert_into_set(vm, &mut self.elements, &item)
+    }
+
+    fn remove(&mut self, item: &PyObjectRef, vm: &VirtualMachine) -> PyResult {
+        fn remove(
+            vm: &VirtualMachine,
+            elements: &mut HashMap<u64, PyObjectRef>,
+            key: u64,
+            value: &PyObjectRef,
+        ) -> PyResult {
+            match elements.remove(&key) {
+                None => {
+                    let item_str = format!("{:?}", value);
+                    Err(vm.new_key_error(item_str))
                 }
+                Some(_) => Ok(vm.get_none()),
             }
-            Err(_) => return Err(vm.new_type_error("".to_string())),
         }
+        perform_action_with_hash(vm, &mut self.elements, &item, &remove)
     }
 
-    Ok(vm.new_bool(false))
-}
+    fn discard(&mut self, item: &PyObjectRef, vm: &VirtualMachine) -> PyResult {
+        fn discard(
+            vm: &VirtualMachine,
+            elements: &mut HashMap<u64, PyObjectRef>,
+            key: u64,
+            _value: &PyObjectRef,
+        ) -> PyResult {
+            elements.remove(&key);
+            Ok(vm.get_none())
+        }
+        perform_action_with_hash(vm, &mut self.elements, &item, &discard)
+    }
 
-fn set_eq(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
-    set_compare_inner(
-        vm,
-        args,
-        &|zelf: usize, other: usize| -> bool { zelf != other },
-        false,
-    )
-}
+    fn clear(&mut self, vm: &VirtualMachine) -> PyResult {
```

Comment:
This doesn't need to return anything, it doesn't raise any exceptions and returning `None` should be implicit.

Reply:
Done